### PR TITLE
Documenting a specific use case where you need a value if and only if it is another key is not present

### PR DESCRIPTION
**Pull request by lemire**
In on-demand, you should visit each field only once, which means that once you have access to the value of a key, you need to consume the content. You cannot just 'save it' and parse it later, you must eagerly parse it even if you may need to throw away the result later.

Fixes https://github.com/simdjson/simdjson/issues/1864

### CODE REVIEW

File: tests/ondemand/ondemand_object_tests.cpp [hunk 1]
```diff
@@ -1074,8 +1074,111 @@ namespace object_tests {
...
+      // corresponding to "a", and later throw the result away.
+      ///////////////////////////////////////////////////////////
+      uint64_t myint;
+      // Notice in the line below that I use chaining... I do not
+      // first convert root["a"] to an object and then seek
+      // "myint" into. Chaining works in this instance because
+      // I only need to access the object 'root["a"]' for a single
+      // key, if I needed to do more work on root["a"], then I should
+      // create an object.
+      if(root["a"]["myint"].get(myint) == SUCCESS) {
```
Comment (strager) [inline tests/ondemand/ondemand_object_tests.cpp:1101]: In my project, the code extracting `myint` is inside `handle_successful_response`, which is a virtual function. Calling `handle_successful_response` before validating the message is a problem. Moving the `myint` extraction into `process` is an abstraction violation.

Your proposed solution is undesirable for my project. I'd prefer a performance hit over an abstraction violation. But if I need to do it this way, I guess I have no choice.

Comment (lemire) [reply tests/ondemand/ondemand_object_tests.cpp:1101]: The following design will not work...

```C++
virtual void handle_successful_response(
    std::uint64_t request_id,
    simdjson::ondemand::value& result) = 0;
```

You need something like this... 

```C++
virtual void handle_successful_response(
    std::uint64_t request_id,
    myresulttype& result) = 0;
```

**Comment by lemire** [comment]
I have improved the documentation somewhat. It is more precise.

File: doc/basics.md [hunk 2]
```diff
@@ -300,10 +304,12 @@ support for users who avoid exceptions. See [the simdjson error handling documen
   comparison. For efficiency reason, you should avoid looking up the same field repeatedly: e.g., do
   not do `object["foo"]` followed by `object["foo"]` with the same `object` instance. If you consume an
   object twice: `std::string_view(object["foo"]` followed by `std::string_view(object["foo"]`, your code
-  is in error. Furthermore, you can only consume one field at a time, on the same object. Thus
-  if you have retrieved `content["bids"].get_array()` and you later call `content["asks"].get_array()`, then the
-  first array should no longer be accessed: it would be unsafe to do so. You can detect such mistakes by first
-  compiling and running the code in Debug mode: an OUT_OF_ORDER_ITERATION error is generated.
+  is in error. Furthermore, you can only consume one field at a time, on the same object. The
+  value instance you get from  `content["bids"]` becomes invalid when you call `content["asks"]`.
```
Comment (strager) [inline doc/basics.md:308]: 👍

**Comment by lemire** [comment]
@strager  I take very seriously your comments and concerns.

**Pull request merged**